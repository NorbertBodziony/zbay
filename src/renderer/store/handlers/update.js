import { ipcRenderer } from 'electron'
import { actionCreators } from './modals'
import { createMigrationFile } from './identity'
export const checkForUpdate = () => async (dispatch, getState) => {
  // create file on new update
  // Note it will recreate file on each new update so file will be up to date
  await dispatch(createMigrationFile())
  dispatch(actionCreators.openModal('applicationUpdate')())
}

export const startApplicationUpdate = () => async (dispatch, getState) => {
  ipcRenderer.send('proceed-update')
  dispatch(actionCreators.closeModal('applicationUpdate')())
}

export const declineUpdate = () => async (dispatch, getState) => {
  dispatch(actionCreators.closeModal('applicationUpdate')())
}

export const epics = {
  checkForUpdate,
  startApplicationUpdate,
  declineUpdate
}

export default {
  epics
}
