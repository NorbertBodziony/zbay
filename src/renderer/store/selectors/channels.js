import { createSelector } from 'reselect'

import zcashChannels from '../../zcash/channels'
import nodeSelectors from './node'

const ignoredChannels = network => [
  zcashChannels.registeredUsers[network].address,
  zcashChannels.channelOfChannels[network].address,
  zcashChannels.priceOracle[network].address
]
const store = s => s
const channels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.filter(
          channel => !ignoredChannels(network).includes(channel.get('address'))
        )
      )
  }
)

const data = createSelector(channels, nodeSelectors.network, (ch, network) => {
  let channelsData = ch.data
  if (channelsData.size) {
    const storeId = channelsData.findIndex(
      ch => ch.get('address') === zcashChannels.store[network].address
    )
    const storeChannel = channelsData.get(storeId)
    channelsData = channelsData.delete(storeId).unshift(storeChannel)
    const zbayId = channelsData.findIndex(
      ch => ch.get('address') === zcashChannels.general[network].address
    )
    const zbayChannel = channelsData.get(zbayId)
    channelsData = channelsData.delete(zbayId).unshift(zbayChannel)
  }
  return channelsData
})
const loader = createSelector(channels, ch => ch.loader)
const errors = createSelector(channels, c => c.get('errors'))

const generalChannelId = createSelector(
  data,
  nodeSelectors.network,
  (ch, network) => {
    const generalChannel = ch.find(
      c => c.get('address') === zcashChannels.general[network].address
    )
    return generalChannel && generalChannel.get('id')
  }
)
const usersChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.registeredUsers[network].address
        )
      ).data
  }
)
const priceOracleChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.priceOracle[network].address
        )
      ).data
  }
)
const publicChannels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.channelOfChannels[network].address
        )
      ).data
  }
)

const channelById = id =>
  createSelector(data, ch => ch.find(c => c.get('id') === id))
const ownedChannels = createSelector(data, ch =>
  ch.filter(c => {
    return c.get('keys').get('sk')
  })
)

const lastSeen = id => createSelector(channelById(id), ch => ch.get('lastSeen'))

export default {
  channelById,
  generalChannelId,
  usersChannel,
  channels,
  loader,
  lastSeen,
  data,
  errors,
  ownedChannels,
  publicChannels,
  priceOracleChannel
}
