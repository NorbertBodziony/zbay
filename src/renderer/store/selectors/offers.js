import { createSelector } from 'reselect'
import * as R from 'ramda'
import directMssagesQueueSelectors from './directMessagesQueue'
import operationsSelectors from './operations'
import removedChannelsSelectors from './removedChannels'
import zbayMessages from '../../zbay/messages'
import identitySelectors from './identity'
import usersSelectors from './users'
import { mergeIntoOne } from './channel'

const store = s => s

const offers = createSelector(
  store,
  s => s.get('offers')
)

const filteredOffers = createSelector(
  store,
  removedChannelsSelectors.removedChannels,
  (s, removedChannels) => {
    const filteredOffers = s
      .get('offers')
      .toList()
      .map((offer, i) => {
        const messages = offer.get('messages')
        const [newestMsg] = messages.sort((a, b) => parseInt(b.createdAt) - parseInt(a.createdAt))
        const removedChannelTimestamp = removedChannels.get(offer.get('itemId'))
        if (!newestMsg) {
          return
        }
        const { createdAt: contactMsgTimestamp } = newestMsg
        if (
          removedChannelTimestamp &&
          parseInt(removedChannelTimestamp) > parseInt(contactMsgTimestamp)
        ) {
          return null
        } else return offer
      })
    return filteredOffers.filter(offer => !R.isNil(offer))
  }
)

const offer = id =>
  createSelector(
    offers,
    a => a.get(id)
  )

export const queuedMessages = id =>
  createSelector(
    directMssagesQueueSelectors.queue,
    queue =>
      queue.filter(
        m => m.offerId === id.substring(0, 64) && m.recipientUsername === id.substring(64)
      )
  )
export const advertMessage = id =>
  createSelector(
    offer(id),
    offer => offer.messages.find(msg => msg.type === 2)
  )

export const pendingMessages = id =>
  createSelector(
    operationsSelectors.operations,
    operations =>
      operations.filter(
        o => o.meta.offerId === id.substring(0, 64) && o.meta.recipientUsername === id.substring(64)
      )
  )
const offerMessages = (id, signerPubKey) =>
  createSelector(
    identitySelectors.data,
    usersSelectors.registeredUser(signerPubKey),
    offers,
    pendingMessages(id),
    queuedMessages(id),
    (identity, registeredUser, a, pendingMessages, queuedMessages) => {
      const userData = registeredUser ? registeredUser.toJS() : null
      const identityAddress = identity.address
      const identityName = userData ? userData.nickname : identity.name
      const displayablePending = pendingMessages.map(operation => {
        return zbayMessages.operationToDisplayableMessage({
          operation: operation.updateIn(['meta', 'message', 'message'], msg => msg.get('text')),
          tag: operation.getIn(['meta', 'message', 'message', 'tag']),
          offerOwner: operation.getIn(['meta', 'message', 'message', 'offerOwner']),
          identityAddress,
          identityName,
          receiver: {
            replyTo: operation.meta.recipientAddress,
            username: operation.meta.recipientUsername
          }
        })
      })
      const displayableQueued = queuedMessages.map((queuedMessage, messageKey) => {
        return zbayMessages.queuedToDisplayableMessage({
          queuedMessage: queuedMessage.updateIn(['message', 'message'], msg => msg.get('text')),
          tag: queuedMessage.getIn(['message', 'message', 'tag']),
          offerOwner: queuedMessage.getIn(['message', 'message', 'offerOwner']),
          messageKey,
          identityAddress,
          identityName,
          receiver: {
            replyTo: queuedMessage.recipientAddress,
            username: queuedMessage.recipientUsername
          }
        })
      })
      const vaultMessages = a.get(id).messages
      const concatedMessages = vaultMessages
        .concat(displayableQueued.values(), displayablePending.values())
        .sortBy(m => m.get('createdAt'))
      const merged = mergeIntoOne(concatedMessages)
      return merged
    }
  )
const lastSeen = id =>
  createSelector(
    offer(id),
    ch => ch.get('lastSeen')
  )
const newMessages = id =>
  createSelector(
    offer(id),
    ch => ch.get('newMessages')
  )
export default {
  offers,
  offer,
  filteredOffers,
  offerMessages,
  lastSeen,
  advertMessage,
  newMessages
}
