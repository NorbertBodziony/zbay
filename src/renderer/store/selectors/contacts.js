import { createSelector } from 'reselect'
import Immutable from 'immutable'
import identitySelectors from './identity'
import directMssagesQueueSelectors from './directMessagesQueue'
import zbayMessages from '../../zbay/messages'
import operationsSelectors from './operations'
import { operationTypes } from '../handlers/operations'
import usersSelectors from './users'
import nodeSelectors from './node'
import { mergeIntoOne } from './channel'

export const Contact = Immutable.Record({
  lastSeen: null,
  username: '',
  address: '',
  messages: Immutable.List(),
  newMessages: Immutable.List(),
  vaultMessages: Immutable.List()
})

const store = s => s

const contacts = createSelector(
  store,
  state => state.get('contacts')
)
const contact = address =>
  createSelector(
    contacts,
    c => c.get(address, Contact())
  )
const messages = address =>
  createSelector(
    contact(address),
    c => c.messages
  )
const allMessages =
  createSelector(
    contacts,
    c => Array.from(c.values()).reduce((acc, contact) => (acc.concat(contact.messages)), Immutable.List())
  )
const pendingBalance =
  createSelector(
    allMessages, nodeSelectors.currentBlock,
    (allMessages, currentBlock) => allMessages.filter(txn => currentBlock - txn.blockTime).reduce((acc, txn) => acc + parseFloat(txn.spent), 0)
  )
const lastSeen = address =>
  createSelector(
    contact(address),
    c => c.lastSeen
  )
const username = address =>
  createSelector(
    contact(address),
    c => c.username
  )
const vaultMessages = address =>
  createSelector(
    contact(address),
    c => c.vaultMessages
  )
const newMessages = address =>
  createSelector(
    contact(address),
    c => c.newMessages
  )

export const queuedMessages = address =>
  createSelector(
    directMssagesQueueSelectors.queue,
    queue => queue.filter(m => m.recipientAddress === address && m.message.get('type') < 10) //  separate offer messages and direct messages
  )

export const pendingMessages = address =>
  createSelector(
    operationsSelectors.operations,
    operations =>
      operations.filter(
        o =>
          o.type === operationTypes.pendingDirectMessage &&
          o.meta.recipientAddress === address &&
          o.meta.message.get('type') < 10 //  separate offer messages and direct messages
      )
  )

export const directMessages = (address, signerPubKey) =>
  createSelector(
    identitySelectors.data,
    usersSelectors.registeredUser(signerPubKey),
    messages(address),
    vaultMessages(address),
    pendingMessages(address),
    queuedMessages(address),
    (identity, registeredUser, messages, vaultMessages, pendingMessages, queuedMessages) => {
      const userData = registeredUser ? registeredUser.toJS() : null
      const identityAddress = identity.address
      const identityName = userData ? userData.nickname : identity.name
      const displayablePending = pendingMessages.map(operation =>
        zbayMessages.operationToDisplayableMessage({
          operation,
          identityAddress,
          identityName,
          receiver: {
            replyTo: operation.meta.recipientAddress,
            username: operation.meta.recipientUsername
          }
        })
      )

      const displayableQueued = queuedMessages.map((queuedMessage, messageKey) =>
        zbayMessages.queuedToDisplayableMessage({
          queuedMessage,
          messageKey,
          identityAddress,
          identityName
        })
      )

      const fetchedMessagesToDisplay = messages.map(msg =>
        zbayMessages.receivedToDisplayableMessage({
          message: msg,
          identityAddress,
          receiver: { replyTo: identityAddress, username: identityName }
        })
      )

      const concatedMessages = fetchedMessagesToDisplay
        .concat(vaultMessages.values(), displayableQueued.values(), displayablePending.values())
        .sortBy(m => m.get('createdAt'))
      const merged = mergeIntoOne(concatedMessages)
      return merged
    }
  )

export default {
  contacts,
  queuedMessages,
  pendingMessages,
  contact,
  messages,
  directMessages,
  lastSeen,
  vaultMessages,
  username,
  newMessages,
  allMessages,
  pendingBalance
}
