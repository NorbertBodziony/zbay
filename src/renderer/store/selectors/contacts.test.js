/* eslint import/first: 0 */
import Immutable from 'immutable'
import * as R from 'ramda'

import create from '../create'
import selectors, { Contact } from './contacts'
import testUtils from '../../testUtils'
import { ReceivedMessage } from '../handlers/messages'
import { operationTypes, PendingDirectMessageOp, Operation } from '../handlers/operations'
import { PendingMessage } from '../handlers/directMessagesQueue'
import { NodeState } from '../handlers/node'

describe('operations selectors', () => {
  const [identity1, identity2] = testUtils.identities
  const messages = Immutable.List(
    R.range(0, 2).map(id => ReceivedMessage(
      testUtils.messages.createReceivedMessage({
        id,
        createdAt: testUtils.now.minus({ hours: 2 * id }).toSeconds(),
        sender: identity1,
        blockTime: 100
      })
    ))
  )
  const vaultMessages = Immutable.List(
    R.range(0, 2).map(id => ReceivedMessage(
      testUtils.messages.createReceivedMessage({
        id,
        createdAt: testUtils.now.minus({ hours: 2 * id }).toSeconds(),
        sender: identity1
      })
    ))
  )
  const recipientUsername = 'test-recipient-username'

  let store = null
  beforeEach(() => {
    store = create({
      initialState: Immutable.Map({
        node: NodeState({
          currentBlock: 123
        }),
        contacts: Immutable.Map({
          [identity1.address]: Contact({
            username: identity1.username,
            address: identity1.address,
            messages,
            vaultMessages,
            lastSeen: testUtils.now,
            newMessages: Immutable.List([ 1, 2, 3, 4 ])
          }),
          [identity2.address]: Contact({
            username: identity2.username,
            address: identity2.address
          })
        }),
        directMessagesQueue: Immutable.Map({
          'messageHash': PendingMessage({
            recipientAddress: identity1.address,
            recipientUsername,
            message: Immutable.fromJS(
              testUtils.createMessage('test-pending-message', testUtils.now.minus({ hours: 2 }).toSeconds())
            )
          })
        }),
        operations: Immutable.fromJS({
          'test-operation-id': Operation({
            opId: 'test-operation-id',
            txId: 'transaction-id',
            type: operationTypes.pendingDirectMessage,
            meta: PendingDirectMessageOp({
              message: Immutable.fromJS(testUtils.createMessage(
                'test-message-id',
                testUtils.now.minus({ hours: 1 }).toSeconds()
              )),
              recipientAddress: identity1.address,
              recipientUsername
            }),
            status: 'success'
          }),
          'test-operation-id-2': Operation({
            opId: 'test-operation-id-2',
            txId: 'transaction-id-2',
            type: operationTypes.pendingDirectMessage,
            meta: PendingDirectMessageOp({
              message: Immutable.fromJS(testUtils.createMessage(
                'test-message-id-2',
                testUtils.now.minus({ hours: 3 }).toSeconds()
              )),
              recipientAddress: identity1.address,
              recipientUsername
            }),
            status: 'success'
          }),
          'test-operation-id-3': Operation({
            opId: 'test-operation-id-3',
            txId: 'transaction-id-3',
            type: operationTypes.pendingDirectMessage,
            meta: PendingDirectMessageOp({
              message: Immutable.fromJS(testUtils.createMessage(
                'test-message-id-3',
                testUtils.now.minus({ hours: 5 }).toSeconds()
              )),
              recipientAddress: identity1.address,
              recipientUsername
            }),
            status: 'success'
          })
        })
      })
    })
    jest.clearAllMocks()
  })

  it(' - contacts', () => {
    expect(selectors.contacts(store.getState())).toMatchSnapshot()
  })

  it(' - contact', () => {
    expect(selectors.contact(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - allMessages', () => {
    expect(selectors.allMessages(store.getState())).toMatchSnapshot()
  })
  it.only(' - pendingBalance', () => {
    expect(selectors.pendingBalance(store.getState())).toMatchSnapshot()
  })
  it(' - messages', () => {
    expect(selectors.messages(identity1.address)(store.getState())).toMatchSnapshot()
  })
  it(' - newMessages', () => {
    expect(selectors.newMessages(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - lastSeen', () => {
    expect(selectors.lastSeen(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - pending messages', () => {
    expect(selectors.pendingMessages(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - queue messages', () => {
    expect(selectors.queuedMessages(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - vault messages', () => {
    expect(selectors.vaultMessages(identity1.address)(store.getState())).toMatchSnapshot()
  })

  it(' - direct messages', () => {
    expect(selectors.directMessages(identity1.address)(store.getState())).toMatchSnapshot()
  })
})
