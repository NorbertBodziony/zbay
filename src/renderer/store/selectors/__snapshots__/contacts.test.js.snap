// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operations selectors  - allMessages 1`] = `
Immutable.List [
  Immutable.ReceivedMessage {
    "id": 0,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551963828,
    "message": "This is a message with id 0",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.ReceivedMessage {
    "id": 1,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551956628,
    "message": "This is a message with id 1",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
]
`;

exports[`operations selectors  - contact 1`] = `
Immutable.Record {
  "lastSeen": "2019-03-07T13:03:48.000Z",
  "username": "Wenus",
  "address": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
  "messages": Immutable.List [
    Immutable.ReceivedMessage {
      "id": 0,
      "type": 1,
      "sender": Immutable.MessageSender {
        "replyTo": "",
        "username": "",
      },
      "createdAt": 1551963828,
      "message": "This is a message with id 0",
      "spent": "0.32",
      "isUnregistered": false,
      "tag": "",
      "offerOwner": "",
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
    Immutable.ReceivedMessage {
      "id": 1,
      "type": 1,
      "sender": Immutable.MessageSender {
        "replyTo": "",
        "username": "",
      },
      "createdAt": 1551956628,
      "message": "This is a message with id 1",
      "spent": "0.32",
      "isUnregistered": false,
      "tag": "",
      "offerOwner": "",
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
  ],
  "newMessages": Immutable.List [
    1,
    2,
    3,
    4,
  ],
  "vaultMessages": Immutable.List [
    Immutable.ReceivedMessage {
      "id": 0,
      "type": 1,
      "sender": Immutable.MessageSender {
        "replyTo": "",
        "username": "",
      },
      "createdAt": 1551963828,
      "message": "This is a message with id 0",
      "spent": "0.32",
      "isUnregistered": false,
      "tag": "",
      "offerOwner": "",
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
    Immutable.ReceivedMessage {
      "id": 1,
      "type": 1,
      "sender": Immutable.MessageSender {
        "replyTo": "",
        "username": "",
      },
      "createdAt": 1551956628,
      "message": "This is a message with id 1",
      "spent": "0.32",
      "isUnregistered": false,
      "tag": "",
      "offerOwner": "",
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
  ],
}
`;

exports[`operations selectors  - contacts 1`] = `
Immutable.Map {
  "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya": Immutable.Record {
    "lastSeen": "2019-03-07T13:03:48.000Z",
    "username": "Wenus",
    "address": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    "messages": Immutable.List [
      Immutable.ReceivedMessage {
        "id": 0,
        "type": 1,
        "sender": Immutable.MessageSender {
          "replyTo": "",
          "username": "",
        },
        "createdAt": 1551963828,
        "message": "This is a message with id 0",
        "spent": "0.32",
        "isUnregistered": false,
        "tag": "",
        "offerOwner": "",
        "publicKey": null,
        "blockTime": 9007199254740991,
      },
      Immutable.ReceivedMessage {
        "id": 1,
        "type": 1,
        "sender": Immutable.MessageSender {
          "replyTo": "",
          "username": "",
        },
        "createdAt": 1551956628,
        "message": "This is a message with id 1",
        "spent": "0.32",
        "isUnregistered": false,
        "tag": "",
        "offerOwner": "",
        "publicKey": null,
        "blockTime": 9007199254740991,
      },
    ],
    "newMessages": Immutable.List [
      1,
      2,
      3,
      4,
    ],
    "vaultMessages": Immutable.List [
      Immutable.ReceivedMessage {
        "id": 0,
        "type": 1,
        "sender": Immutable.MessageSender {
          "replyTo": "",
          "username": "",
        },
        "createdAt": 1551963828,
        "message": "This is a message with id 0",
        "spent": "0.32",
        "isUnregistered": false,
        "tag": "",
        "offerOwner": "",
        "publicKey": null,
        "blockTime": 9007199254740991,
      },
      Immutable.ReceivedMessage {
        "id": 1,
        "type": 1,
        "sender": Immutable.MessageSender {
          "replyTo": "",
          "username": "",
        },
        "createdAt": 1551956628,
        "message": "This is a message with id 1",
        "spent": "0.32",
        "isUnregistered": false,
        "tag": "",
        "offerOwner": "",
        "publicKey": null,
        "blockTime": 9007199254740991,
      },
    ],
  },
  "zs1dhqp9dtr4pksnmaynp2k22qduvywejg3neqq4swd4a6gnz6w0m208kefcdm9n2067yn5clcvgsq": Immutable.Record {
    "lastSeen": null,
    "username": "Mars",
    "address": "zs1dhqp9dtr4pksnmaynp2k22qduvywejg3neqq4swd4a6gnz6w0m208kefcdm9n2067yn5clcvgsq",
    "messages": Immutable.List [],
    "newMessages": Immutable.List [],
    "vaultMessages": Immutable.List [],
  },
}
`;

exports[`operations selectors  - direct messages 1`] = `
Immutable.List [
  Immutable.DisplayableMessage {
    "id": "test-operation-id-3",
    "type": 1,
    "sender": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "receiver": Immutable.ExchangeParticipant {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "username": "test-recipient-username",
    },
    "createdAt": 1551945828,
    "message": "This is some message test-message-id-3",
    "spent": "0",
    "fromYou": true,
    "status": "success",
    "error": null,
    "shippingData": null,
    "tag": "",
    "offerOwner": null,
    "isUnregistered": false,
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.DisplayableMessage {
    "id": "test-operation-id-2",
    "type": 1,
    "sender": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "receiver": Immutable.ExchangeParticipant {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "username": "test-recipient-username",
    },
    "createdAt": 1551953028,
    "message": "This is some message test-message-id-2",
    "spent": "0",
    "fromYou": true,
    "status": "success",
    "error": null,
    "shippingData": null,
    "tag": "",
    "offerOwner": null,
    "isUnregistered": false,
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.DisplayableMessage {
    "id": 1,
    "type": 1,
    "sender": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "receiver": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551956628,
    "message": "This is a message with id 1
This is a message with id 1
This is some message test-pending-message",
    "spent": "0.32",
    "fromYou": true,
    "status": "pending",
    "error": null,
    "shippingData": null,
    "tag": "",
    "offerOwner": "",
    "isUnregistered": false,
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.DisplayableMessage {
    "id": "test-operation-id",
    "type": 1,
    "sender": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "receiver": Immutable.ExchangeParticipant {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "username": "test-recipient-username",
    },
    "createdAt": 1551960228,
    "message": "This is some message test-message-id",
    "spent": "0",
    "fromYou": true,
    "status": "success",
    "error": null,
    "shippingData": null,
    "tag": "",
    "offerOwner": null,
    "isUnregistered": false,
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.DisplayableMessage {
    "id": 0,
    "type": 1,
    "sender": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "receiver": Immutable.ExchangeParticipant {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551963828,
    "message": "This is a message with id 0
This is a message with id 0",
    "spent": "0.32",
    "fromYou": true,
    "status": "broadcasted",
    "error": null,
    "shippingData": null,
    "tag": "",
    "offerOwner": "",
    "isUnregistered": false,
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
]
`;

exports[`operations selectors  - lastSeen 1`] = `"2019-03-07T13:03:48.000Z"`;

exports[`operations selectors  - messages 1`] = `
Immutable.List [
  Immutable.ReceivedMessage {
    "id": 0,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551963828,
    "message": "This is a message with id 0",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.ReceivedMessage {
    "id": 1,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551956628,
    "message": "This is a message with id 1",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
]
`;

exports[`operations selectors  - newMessages 1`] = `
Immutable.List [
  1,
  2,
  3,
  4,
]
`;

exports[`operations selectors  - pending messages 1`] = `
Immutable.Map {
  "test-operation-id": Immutable.Operation {
    "opId": "test-operation-id",
    "type": "pendingDirectMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551960228,
        "message": "This is some message test-message-id",
        "signature": Object {
          "data": Array [
            202,
            15,
            119,
            117,
            178,
            163,
            37,
            253,
            154,
            252,
            231,
            166,
            195,
            137,
            183,
            234,
            71,
            243,
            80,
            122,
            59,
            238,
            222,
            166,
            186,
            69,
            90,
            160,
            160,
            50,
            104,
            190,
            10,
            207,
            36,
            153,
            83,
            108,
            75,
            44,
            43,
            251,
            163,
            248,
            121,
            50,
            116,
            127,
            154,
            101,
            218,
            181,
            222,
            56,
            194,
            169,
            79,
            74,
            219,
            181,
            87,
            23,
            91,
            207,
          ],
          "type": "Buffer",
        },
      },
      "recipientAddress": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "recipientUsername": "test-recipient-username",
      "offerId": "",
    },
    "txId": "transaction-id",
    "error": null,
    "status": "success",
  },
  "test-operation-id-2": Immutable.Operation {
    "opId": "test-operation-id-2",
    "type": "pendingDirectMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id-2",
        "type": 1,
        "createdAt": 1551953028,
        "message": "This is some message test-message-id-2",
        "signature": Object {
          "data": Array [
            93,
            229,
            160,
            223,
            47,
            10,
            65,
            174,
            71,
            112,
            184,
            230,
            247,
            84,
            67,
            130,
            157,
            4,
            33,
            195,
            30,
            65,
            144,
            47,
            131,
            90,
            154,
            213,
            24,
            180,
            246,
            234,
            127,
            112,
            70,
            23,
            253,
            17,
            207,
            91,
            184,
            206,
            3,
            81,
            254,
            9,
            112,
            222,
            134,
            123,
            230,
            112,
            178,
            116,
            203,
            168,
            175,
            240,
            4,
            241,
            69,
            17,
            27,
            101,
          ],
          "type": "Buffer",
        },
      },
      "recipientAddress": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "recipientUsername": "test-recipient-username",
      "offerId": "",
    },
    "txId": "transaction-id-2",
    "error": null,
    "status": "success",
  },
  "test-operation-id-3": Immutable.Operation {
    "opId": "test-operation-id-3",
    "type": "pendingDirectMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id-3",
        "type": 1,
        "createdAt": 1551945828,
        "message": "This is some message test-message-id-3",
        "signature": Object {
          "data": Array [
            239,
            184,
            201,
            81,
            172,
            7,
            154,
            148,
            230,
            112,
            241,
            11,
            130,
            31,
            77,
            124,
            79,
            31,
            73,
            239,
            138,
            184,
            107,
            171,
            230,
            45,
            49,
            68,
            107,
            110,
            89,
            216,
            62,
            27,
            132,
            250,
            112,
            76,
            34,
            1,
            215,
            116,
            57,
            162,
            94,
            114,
            218,
            56,
            166,
            48,
            45,
            91,
            225,
            213,
            151,
            95,
            17,
            38,
            129,
            187,
            248,
            87,
            234,
            10,
          ],
          "type": "Buffer",
        },
      },
      "recipientAddress": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
      "recipientUsername": "test-recipient-username",
      "offerId": "",
    },
    "txId": "transaction-id-3",
    "error": null,
    "status": "success",
  },
}
`;

exports[`operations selectors  - pendingBalance 1`] = `0.64`;

exports[`operations selectors  - queue messages 1`] = `
Immutable.Map {
  "messageHash": Immutable.PendingDirectMessage {
    "recipientAddress": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    "recipientUsername": "test-recipient-username",
    "offerId": "",
    "message": Immutable.Map {
      "id": "test-pending-message",
      "type": 1,
      "createdAt": 1551956628,
      "message": "This is some message test-pending-message",
      "signature": Object {
        "data": Array [
          125,
          48,
          0,
          208,
          24,
          4,
          145,
          148,
          20,
          58,
          51,
          195,
          92,
          109,
          173,
          90,
          216,
          205,
          238,
          114,
          236,
          220,
          98,
          151,
          54,
          136,
          21,
          232,
          114,
          84,
          38,
          225,
          12,
          47,
          69,
          14,
          21,
          122,
          217,
          230,
          156,
          174,
          99,
          112,
          233,
          199,
          79,
          156,
          18,
          29,
          223,
          61,
          38,
          174,
          58,
          169,
          170,
          87,
          77,
          246,
          158,
          219,
          15,
          48,
        ],
        "type": "Buffer",
      },
    },
  },
}
`;

exports[`operations selectors  - vault messages 1`] = `
Immutable.List [
  Immutable.ReceivedMessage {
    "id": 0,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551963828,
    "message": "This is a message with id 0",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
  Immutable.ReceivedMessage {
    "id": 1,
    "type": 1,
    "sender": Immutable.MessageSender {
      "replyTo": "",
      "username": "",
    },
    "createdAt": 1551956628,
    "message": "This is a message with id 1",
    "spent": "0.32",
    "isUnregistered": false,
    "tag": "",
    "offerOwner": "",
    "publicKey": null,
    "blockTime": 9007199254740991,
  },
]
`;
